File: php/src/Exception/FivexerApiException.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Exception;

use Fivexer\SDK\Model\QuotaInfo;

/**
 * Raised when the /v1 API returns a non-2xx response.
 *
 * Mirrors the platform error envelope { "error": { "code", "message" } }.
 * `$apiCode` carries the API's error code (e.g. rate_limited, plan_limit_exceeded,
 * not_found, task_exists, validation_failed) — not the inherited getCode(), which is
 * PHP's numeric exception code and is always 0 here. On 402/429 the quota snapshot at
 * the moment of rejection is attached, plus retry-after.
 */
class FivexerApiException extends FivexerException
{
    /**
     * @param int $statusCode HTTP status code returned by the API
     * @param string $apiCode API error code (e.g. 'rate_limited')
     * @param string $message Human-readable error message
     * @param QuotaInfo|null $quota Quota state at the moment of rejection (402/429)
     * @param float|null $retryAfterSeconds Value of Retry-After (seconds) when present
     */
    public function __construct(
        public readonly int $statusCode,
        public readonly string $apiCode,
        string $message,
        public readonly ?QuotaInfo $quota = null,
        public readonly ?float $retryAfterSeconds = null,
    ) {
        parent::__construct($message);
    }

    public function __toString(): string
    {
        return 'FivexerApiException{statusCode=' . $this->statusCode
            . ', apiCode=\'' . $this->apiCode . '\', message=\'' . $this->getMessage() . '\'}';
    }
}

File: java/src/main/java/io/fivexer/sdk/FivexerApiException.java
package io.fivexer.sdk;

import io.fivexer.sdk.model.QuotaInfo;

/**
 * Raised when the /v1 API returns a non-2xx response.
 *
 * <p>Mirrors the platform error envelope {@code { "error": { "code", "message" } }}.
 * The {@link #getCode()} matches the API's error codes (e.g. {@code rate_limited},
 * {@code plan_limit_exceeded}, {@code not_found}, {@code task_exists},
 * {@code validation_failed}). On {@code 402}/{@code 429} the quota snapshot at the moment
 * of rejection is attached, plus {@link #getRetryAfterSeconds()} when present.
 */
public class FivexerApiException extends FivexerException {
    private static final long serialVersionUID = 1L;

    private final int statusCode;
    private final String code;
    private final QuotaInfo quota;
    private final Double retryAfterSeconds;

    public FivexerApiException(int statusCode, String code, String message, QuotaInfo quota, Double retryAfterSeconds) {
        super(message);
        this.statusCode = statusCode;
        this.code = code;
        this.quota = quota;
        this.retryAfterSeconds = retryAfterSeconds;
    }

    public int getStatusCode() {
        return statusCode;
    }

    public String getCode() {
        return code;
    }

    /** Quota state at the moment of rejection (present on 402/429), or {@code null}. */
    public QuotaInfo getQuota() {
        return quota;
    }

    /** Value of {@code Retry-After} (seconds) when present, or {@code null}. */
    public Double getRetryAfterSeconds() {
        return retryAfterSeconds;
    }

    @Override
    public String toString() {
        return "FivexerApiException{statusCode=" + statusCode + ", code='" + code + "', message='" + getMessage() + "'}";
    }
}
